Inspection numbers linked to formal charges are listed at the end of this page.

In [2]:
## So first let's import the data and set up our libraries 
import pandas as pd 
wm_data = pd.read_excel("Weights-Measures_Data.xlsx")

In [3]:
##Great now let's take a peek under the hood of the data 
wm_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12755 entries, 0 to 12754
Data columns (total 35 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Certificate of Inspection      12755 non-null  object        
 1   Inspection Number              12755 non-null  object        
 2   Date of Occurance              12755 non-null  datetime64[ns]
 3   Year (added)                   12755 non-null  int64         
 4   Business Unique ID             12750 non-null  object        
 5   Business Name                  12755 non-null  object        
 6   DBA Trade Name                 3552 non-null   object        
 7   Business Category              12755 non-null  object        
 8   Inspection type                12755 non-null  object        
 9   Device Category                12755 non-null  object        
 10  Device Class                   12699 non-null  object        
 11  Device Type    

In [4]:
## First we want to verify if the Inspection Number values are unique
wm_data['Inspection Number'].is_unique
pd.set_option('display.max_columns', None)  
wm_data[wm_data['Inspection Number'].duplicated(keep=False)]

,Certificate of Inspection,Inspection Number,Date of Occurance,Year (added),Business Unique ID,Business Name,DBA Trade Name,Business Category,Inspection type,Device Category,Device Class,Device Type,Devices Inspected,Devices Approved,Devices Condemned/Confiscated,Confiscated,Address Type,Building Number,Street Name,Street Name 2,Unit Type,Unit,City,State,Postcode,Borough,Community Board,Council District,BIN,BBL,NTA,Latitude,Longitude,,.1
5,A0156217,INS-000720678,2024-02-28,2024,BA-1559330-2022,THE HEIGHTS MARKET CORP,NaN,Grocery-Retail - 808,Scale,Scale,Scale Inspection,Computing Scale - 01,4,4,0,NaN,Complete Address,1520,SAINT NICHOLAS AVE,NaN,NaN,NaN,NEW YORK,NY,10033,Manhattan,112.0,10.0,1063788.0,1.021570e+09,MN35,40.852479,-73.931411,NaN,NaN
6,A0156217,INS-000720678,2024-02-28,2024,BA-1559330-2022,THE HEIGHTS MARKET CORP,NaN,Grocery-Retail - 808,Scale,Scale,Scale Inspection,Pre-pack Scale - 01,1,1,0,NaN,Complete Address,1520,SAINT NICHOLAS AVE,NaN,NaN,NaN,NEW YORK,NY,10033,Manhattan,112.0,10.0,1063788.0,1.021570e+09,MN35,40.852479,-73.931411,NaN,NaN
17,A0156705,INS-000721752,2024-03-04,2024,BA-1088978-2022,A & C SEAFOOD & MEAT MARKET INC.,NaN,Supermarket - 819,Scale,Scale,Scale Inspection,Computing Scale - 01,5,5,0,NaN,Complete Address,1237,AVENUE U,NaN,NaN,NaN,BROOKLYN,NY,11229,Brooklyn,315.0,48.0,3197984.0,3.073170e+09,BK25,40.598558,-73.958956,NaN,NaN
18,A0156705,INS-000721752,2024-03-04,2024,BA-1088978-2022,A & C SEAFOOD & MEAT MARKET INC.,NaN,Supermarket - 819,Scale,Scale,Scale Inspection,Customer Scale - 01,1,1,0,NaN,Complete Address,1237,AVENUE U,NaN,NaN,NaN,BROOKLYN,NY,11229,Brooklyn,315.0,48.0,3197984.0,3.073170e+09,BK25,40.598558,-73.958956,NaN,NaN
19,A0156705,INS-000721752,2024-03-04,2024,BA-1088978-2022,A & C SEAFOOD & MEAT MARKET INC.,NaN,Supermarket - 819,Scale,Scale,Scale Inspection,Pre-pack Scale - 01,6,6,0,NaN,Complete Address,1237,AVENUE U,NaN,NaN,NaN,BROOKLYN,NY,11229,Brooklyn,315.0,48.0,3197984.0,3.073170e+09,BK25,40.598558,-73.958956,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12737,A0205858,INS-000816521,2025-01-02,2025,BA-1726866-2025,Good fortune supermarket,NaN,Supermarket - 819,Scale,Scale,Scale Inspection,Pre-pack Scale - 01,2,2,0,NaN,Complete Address,4276,MAIN ST,NaN,NaN,NaN,FLUSHING,NY,11355,Queens,407.0,20.0,4622994.0,4.051248e+09,QN22,40.753137,-73.827283,NaN,NaN
12740,A0211733,INS-000829532,2025-02-12,2025,BA-1727176-2025,LINA FOOD CORP.,C-Town Supermarket,Supermarket - 819,Scale,Scale,Scale Inspection,Computing Scale - 01,9,9,0,NaN,Complete Address,4918,BROADWAY,NaN,NaN,NaN,NEW YORK,NY,10034,Manhattan,112.0,10.0,1082052.0,1.022350e+09,MN01,40.867432,-73.921919,NaN,NaN
12741,A0214071,INS-000833234,2025-02-28,2025,BA-1336537-2022,GREENBAY ORGANIC FOODS LLC,NaN,Grocery-Retail - 808,Scale,Scale,Scale Inspection,Pre-pack Scale - 01,1,1,0,NaN,Complete Address,3114,30TH AVE,NaN,NaN,NaN,ASTORIA,NY,11102,Queens,401.0,22.0,4008488.0,4.006150e+09,QN70,40.766788,-73.921196,NaN,NaN
12748,A0216338,INS-000837260,2025-03-14,2025,BA-1192106-2022,"CHEF'S CHOICE CASH & CARRY FOOD DISTRIBUTOR, INC.",NaN,Wholesale Food Market - 718,Scale,Scale,Scale Inspection,Computing Scale - 01,4,4,0,NaN,Complete Address,1051,UTICA AVE,NaN,NaN,NaN,BROOKLYN,NY,11203,Brooklyn,317.0,45.0,3104354.0,3.047350e+09,BK96,40.647234,-73.929888,NaN,NaN


In [5]:
## We know that the inspection number values are not unique. There can be one inspection number used for multiple devices.
# Create standardized Full Address
wm_data["Full Address"] = (
    wm_data["Building Number"].fillna('').astype(str).str.strip() + " " +
    wm_data["Street Name"].fillna('').astype(str).str.strip() + ", " +
    wm_data["Postcode"].astype(str)
)

# Drop duplicates based on Inspection Number to avoid double-counting
unique_inspections = wm_data.drop_duplicates(subset=["Inspection Number"])

# Count top addresses by number of unique inspections
top_addresses = unique_inspections["Full Address"].value_counts().head(10)

# Show result
top_addresses

Full Address
245 MESEROLE AVE, 11222           943
 CENTRAL PARK WEST, 10023          68
245 MESEROLE AVENUE, 11222         33
 EAST 17 STREET, 10003             21
SEC COLUMBUS AVENUE, 10024          8
880 GARRISON AVE, 10474             8
1317 OAKPOINT AVE, 10474            7
1 BAY BLVD, 11559                   7
1 JFK INTL AIRPORT BLVD, 11430      7
20 GRAND ARMY PLZ, 11238            7
Name: count, dtype: int64

In [6]:
## Okay...so now let's tackle these 7 key investigative questions: 
## 1. Identify the repeat offenders: businesses with multiple failed inspections or confiscated/condemned devices.
## 2. Identify the geographic hotspots of device violations by borough and postal code 
## 3. Identify the most commonly inspected and/or condemned device type 
## 4. Are there seasonal spikes or trends in failures or enforcement
## 5. What business types are most at risk of violations 
## 6. Which borough and business types received the highest confiscations
## 7. How likely are devices to be condemned based on routine vs. complaint-driven inspections.

In [7]:
## Let's first identify the repeat offenders: businesses with multiple failed inspections or confiscated/condemned devices.
## Let's first filter for failed inspections
## Then we'll count the number of failed inspections and look at the top 10

# Step 1: Filter to rows where any devices were condemned or confiscated
failed_inspections = wm_data[wm_data["Devices Condemned/Confiscated"] > 0]

# Step 2: Drop duplicates by Inspection Number to count each failed inspection only once
unique_failed = failed_inspections.drop_duplicates(subset=["Inspection Number"])

# Step 3: Count how many unique failed inspections each business had
repeat_offenders = unique_failed["Business Name"].value_counts()

# Step 4: Show top 10 repeat offenders
repeat_offenders.head(10)

Business Name
BOLLA OPERATING CORP.              83
SPEEDWAY LLC                       43
AMG RETAIL I LLC                   16
SPRAGUE OPERATING RESOURCES LLC    13
JAG TRANSPORTING INC               12
LOGIC FUEL TRANSPORTATION CORP      8
STAR ENERGY TRANSPORTATION INC.     7
E.S.F. TRANSPORT INC.               6
FIVE BORO FUEL OIL LTD              5
CERTIFIED HEATING OILS INC.         4
Name: count, dtype: int64

In [8]:
## These are the most common device types to fail inspections
# Step 1: Filter rows where devices were condemned or confiscated
failed_inspections = wm_data[wm_data["Devices Condemned/Confiscated"] > 0]

# Step 2: Remove duplicates based on Inspection Number + Device Type
unique_failed_devices = failed_inspections.drop_duplicates(subset=["Inspection Number", "Device Type"])

# Step 3: Count the most common device types to fail
most_common_devices = unique_failed_devices["Device Type"].value_counts().head(10)

# Step 4: Display results
most_common_devices

Device Type
Octane Sample           1135
Petrol Meter Type B       86
Computing Scale - 01      81
Petrol Meter Type A       31
Petrol Pump Blend         14
Pre-pack Scale - 01       14
Petrol Pump Single        12
No Fee Scale              12
Platform Scale - 02        9
Petrol Pump Diesel         8
Name: count, dtype: int64

In [9]:
## let's see the general info 
unique_failed_devices.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1416 entries, 56 to 12731
Data columns (total 36 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   Certificate of Inspection      1416 non-null   object        
 1   Inspection Number              1416 non-null   object        
 2   Date of Occurance              1416 non-null   datetime64[ns]
 3   Year (added)                   1416 non-null   int64         
 4   Business Unique ID             1416 non-null   object        
 5   Business Name                  1416 non-null   object        
 6   DBA Trade Name                 525 non-null    object        
 7   Business Category              1416 non-null   object        
 8   Inspection type                1416 non-null   object        
 9   Device Category                1416 non-null   object        
 10  Device Class                   1400 non-null   object        
 11  Device Type         

In [10]:
# Now from our list of top repeat offenders, let's see what business category they fall into
# First, extract the list of top repeat offenders
top_offenders = [
    "BOLLA OPERATING CORP.",
    "SPEEDWAY LLC",
    "AMG RETAIL I LLC",
    "SPRAGUE OPERATING RESOURCES LLC",
    "JAG TRANSPORTING INC",
    "LOGIC FUEL TRANSPORTATION CORP",
    "STAR ENERGY TRANSPORTATION INC.",
    "E.S.F. TRANSPORT INC.",
    "BERIL CORP",
    "FIVE BORO FUEL OIL LTD"
]

# Step 1: Use unique failed inspections (one row per inspection)
unique_failed = wm_data[wm_data["Devices Condemned/Confiscated"] > 0].drop_duplicates(subset=["Inspection Number"])

# Step 2: Filter for just the top repeat offenders
top_failed = unique_failed[unique_failed["Business Name"].isin(top_offenders)]

# Step 3: Check distinct combinations of business name and category
top_failed[["Business Name", "Business Category"]].drop_duplicates()

,Business Name,Business Category
456,SPRAGUE OPERATING RESOURCES LLC,Fuel Oil Dealer - 814
541,BOLLA OPERATING CORP.,Gas Station-Retail - 815
3140,SPEEDWAY LLC,Gas Station-Retail - 815
5234,BERIL CORP,Gas Station-Retail - 815
5253,STAR ENERGY TRANSPORTATION INC.,Fuel Oil Dealer - 814
5337,AMG RETAIL I LLC,Gas Station-Retail - 815
5689,JAG TRANSPORTING INC,Fuel Oil Dealer - 814
5754,LOGIC FUEL TRANSPORTATION CORP,Fuel Oil Dealer - 814
5944,SPRAGUE OPERATING RESOURCES LLC,Gasoline Truck-Retail - 822
6063,E.S.F. TRANSPORT INC.,Fuel Oil Dealer - 814


In [11]:
## Now lets get a ranked list of what devices were being condemned at each repeat offender's inspection
# Use full failed inspections (per device row)
failed_inspections = wm_data[wm_data["Devices Condemned/Confiscated"] > 0]

# Filter to just top offenders
top_failed_devices = failed_inspections[failed_inspections["Business Name"].isin(top_offenders)]

# Count device types condemned at these businesses
device_counts = top_failed_devices["Device Type"].value_counts()

# Display ranked list
device_counts

Device Type
Octane Sample          140
Petrol Meter Type B     49
Petrol Pump Diesel       5
Petrol Pump Blend        4
No Fee Gas Pump          2
Petrol Meter Type A      2
Petrol Pump Single       1
Name: count, dtype: int64

In [12]:
##Now let's see which device type had the most failure per business
# Full failed inspections at device level
failed_inspections = wm_data[wm_data["Devices Condemned/Confiscated"] > 0]

# Filter for top offenders
top_failed_devices = failed_inspections[failed_inspections["Business Name"].isin(top_offenders)]

# Group by Business and Device Type, count failures
device_failures_per_business = top_failed_devices.groupby(
    ["Business Name", "Device Type"]
).size().sort_values(ascending=False)

device_failures_per_business

Business Name                    Device Type        
BOLLA OPERATING CORP.            Octane Sample          82
SPEEDWAY LLC                     Octane Sample          38
AMG RETAIL I LLC                 Octane Sample          16
SPRAGUE OPERATING RESOURCES LLC  Petrol Meter Type B    13
JAG TRANSPORTING INC             Petrol Meter Type B    12
LOGIC FUEL TRANSPORTATION CORP   Petrol Meter Type B     7
STAR ENERGY TRANSPORTATION INC.  Petrol Meter Type B     7
E.S.F. TRANSPORT INC.            Petrol Meter Type B     6
FIVE BORO FUEL OIL LTD           Petrol Meter Type B     4
BERIL CORP                       Octane Sample           4
SPEEDWAY LLC                     Petrol Pump Diesel      3
                                 No Fee Gas Pump         2
BOLLA OPERATING CORP.            Petrol Pump Blend       2
SPEEDWAY LLC                     Petrol Pump Blend       2
BOLLA OPERATING CORP.            Petrol Pump Diesel      1
FIVE BORO FUEL OIL LTD           Petrol Meter Type A     1
LOG

Between 2023 and 2025, Bolla Operating Corp. topped the list of repeat offenders in NYC’s Weights and Measures inspections, failing 82 inspections across multiple locations — more than any other company. Speedway LLC followed with 38 failed inspections, and AMG Retail I LLC with 16.

The violations uncovered go beyond just faulty gas pumps. At the heart of these inspections were two critical issues: fuel quality and metering accuracy — both of which directly impact how much fuel consumers receive and what they pay for.

Bolla alone failed 82 octane sample tests, the highest in the city, suggesting a pattern of mislabeled or tampered fuel — a practice that could mean consumers are paying premium prices for regular-grade gas. In total, the top 10 offenders were cited in over 140 octane sampling failures, pointing to a widespread and systemic issue in fuel integrity across NYC.

At the same time, wholesalers like Sprague Operating Resources and JAG Transporting Inc. were repeatedly cited for Petrol Meter Type B violations — truck-mounted flow meters used to deliver heating oil or diesel to buildings and commercial customers. These meters, when miscalibrated, can result in underdelivery or overbilling, raising serious concerns about fuel delivery fraud in the supply chain.

In short, NYC’s most frequently cited fuel businesses failed not only to dispense fuel accurately — they often failed to deliver what they claimed was in the tank at all. These violations reflect more than broken hardware; they point to systemic negligence — or worse, economic deception — affecting everyday consumers and businesses alike.

In [14]:
## 2. Identify the geographic hotspots of device violations by borough, address, and postal code
filtered_inspections = wm_data[wm_data["Devices Condemned/Confiscated"] > 0]
## Now let's filter the top boroughs 
# Drop duplicate inspections (keep one row per inspection)
unique_inspections = filtered_inspections.drop_duplicates(subset=["Inspection Number"])

# Now count boroughs by unique inspections
borough_violations = unique_inspections["Borough"].value_counts()

borough_violations

Borough
Brooklyn         481
Queens           453
Bronx            256
Staten Island    113
Manhattan         85
Outside NYC        3
Name: count, dtype: int64

In [15]:
## Now let's filter the top zip codes
# Drop duplicate inspections based on Inspection Number
unique_inspections = filtered_inspections.drop_duplicates(subset=["Inspection Number"])

# Count zip codes by unique inspections
zip_violations = unique_inspections["Postcode"].value_counts()

zip_violations.head(10)

Postcode
11222    125
11101     31
11236     30
11434     24
11385     24
11207     22
11234     21
11218     20
10462     20
10314     19
Name: count, dtype: int64

In [16]:
## Now let's filter the top business addresses 
# Create 'Full Address' column
filtered_inspections["Full Address"] = (
    filtered_inspections["Building Number"].fillna('').astype(str).str.strip() + " " +
    filtered_inspections["Street Name"].fillna('').astype(str).str.strip() + ", " +
    filtered_inspections["Postcode"].astype(str)
)

# Drop duplicate inspections by Inspection Number
unique_inspections = filtered_inspections.drop_duplicates(subset=["Inspection Number"])

# Count unique inspections per address
address_violations = unique_inspections["Full Address"].value_counts()

address_violations

/var/folders/6p/9zpd5gjx1dj6zpc834rhv6d40000gn/T/ipykernel_45694/4119333716.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_inspections["Full Address"] = (


Full Address
245 MESEROLE AVE, 11222                109
2985 JEROME AVE, 10468                   4
196-03 GRAND CENTRAL PARKWAY, 11423      4
8521 7TH AVE, 11228                      4
117 MORNINGSIDE AVE, 10027               4
                                      ... 
2317 CONEY ISLAND AVE, 11223             1
10020 METROPOLITAN AVE, 11375            1
1810 FOREST AVE, 10303                   1
1650 FOREST AVE, 10302                   1
14615 UNION TURNPIKE, 11367              1
Name: count, Length: 803, dtype: int64

In [17]:
address_violations.to_csv("address_violations_counts.csv", header=True)

In [18]:
## Our research confirms that the centralized testing station is located at 245 Meserole Avenue, Brooklyn, NY 11222
# Step 1: Exclude centralized testing station at 245 Meserole
remaining_inspections = failed_inspections.copy()
remaining_inspections = remaining_inspections[
    ~(
        (remaining_inspections["Building Number"].astype(str).str.strip() == "245") &
        (remaining_inspections["Street Name"].str.upper().str.contains("MESEROLE"))
    )
]

# Step 2: Create 'Full Address' column
remaining_inspections["Full Address"] = (
    remaining_inspections["Building Number"].fillna('').astype(str).str.strip() + " " +
    remaining_inspections["Street Name"].fillna('').astype(str).str.strip() + ", " +
    remaining_inspections["Postcode"].astype(str)
)

# Step 3: Drop duplicate inspections to count unique inspection events
unique_remaining = remaining_inspections.drop_duplicates(subset=["Inspection Number"])

# Step 4: Count addresses by unique inspections
address_counts = unique_remaining["Full Address"].value_counts()

# Step 5: Display top 10 addresses
address_counts.head(10)

Full Address
1314 SEDGWICK AVE, 10452               4
2173 GRAND CONCOURSE, 10453            4
117 MORNINGSIDE AVE, 10027             4
2985 JEROME AVE, 10468                 4
8521 7TH AVE, 11228                    4
914 REMSEN AVE, 11236                  4
1981 OCEAN AVE, 11230                  4
1982 BRONXDALE AVE, 10462              4
1193 MYRTLE AVE, 11221                 4
196-03 GRAND CENTRAL PARKWAY, 11423    4
Name: count, dtype: int64

In [19]:
## Now let's see the most popular device types from the remaining inspections
device_type_counts = remaining_inspections["Device Type"].value_counts()
device_type_counts.head(10)

Device Type
Octane Sample           1135
Computing Scale - 01      82
Petrol Pump Blend         15
Pre-pack Scale - 01       14
Petrol Pump Single        12
No Fee Scale              12
Petrol Pump Diesel         9
Platform Scale - 02        9
No Fee Gas Pump            7
Petrol Meter Type B        5
Name: count, dtype: int64

In [20]:
## Now let’s identify the top boroughs where violations occurred — excluding inspections conducted at the centralized Meserole testing site.
# Ensure all fields are strings for safe comparison
# Step 1: Prepare fields and exclude Meserole site (your code)
filtered_inspections = failed_inspections.copy()

filtered_inspections["Building Number"] = filtered_inspections["Building Number"].astype(str).str.strip()
filtered_inspections["Street Name"] = filtered_inspections["Street Name"].astype(str).str.strip().str.upper()

remaining_inspections = filtered_inspections[
    ~(
        (filtered_inspections["Building Number"] == "245") &
        (filtered_inspections["Street Name"].str.contains("MESEROLE", na=False))
    )
]

# Step 2: Deduplicate by Inspection Number to count unique inspections
unique_remaining = remaining_inspections.drop_duplicates(subset=["Inspection Number"])

# Step 3: Count borough violations by unique inspections
borough_violations = unique_remaining["Borough"].value_counts()

borough_violations

Borough
Queens           453
Brooklyn         370
Bronx            256
Staten Island    113
Manhattan         85
Outside NYC        3
Name: count, dtype: int64

In [21]:
# Count violations by ZIP/postal code
# Drop duplicate inspections first
unique_remaining = remaining_inspections.drop_duplicates(subset=["Inspection Number"])

# Count by Postcode using unique inspections only
postal_violations = unique_remaining["Postcode"].value_counts()

postal_violations.head(10)

Postcode
11101    31
11236    30
11434    24
11385    24
11207    22
11234    21
11218    20
10462    20
11208    19
10314    19
Name: count, dtype: int64

In [22]:
# Create Full Address column
remaining_inspections["Full Address"] = (
    remaining_inspections["Building Number"].fillna('').astype(str).str.strip() + " " +
    remaining_inspections["Street Name"].fillna('').astype(str).str.strip() + ", " +
    remaining_inspections["Postcode"].astype(str)
)

# Drop duplicates by Inspection Number to count each inspection once
unique_remaining = remaining_inspections.drop_duplicates(subset=["Inspection Number"])

# Count addresses by unique inspections
top_addresses = unique_remaining["Full Address"].value_counts()

top_addresses.head(20)

/var/folders/6p/9zpd5gjx1dj6zpc834rhv6d40000gn/T/ipykernel_45694/1534240882.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  remaining_inspections["Full Address"] = (


Full Address
1314 SEDGWICK AVE, 10452               4
2173 GRAND CONCOURSE, 10453            4
117 MORNINGSIDE AVE, 10027             4
2985 JEROME AVE, 10468                 4
8521 7TH AVE, 11228                    4
914 REMSEN AVE, 11236                  4
1981 OCEAN AVE, 11230                  4
1982 BRONXDALE AVE, 10462              4
1193 MYRTLE AVE, 11221                 4
196-03 GRAND CENTRAL PARKWAY, 11423    4
1317 OAKPOINT AVE, 10474               3
2430 FDR DR, 10010                     3
2001 CROPSEY AVE, 11214                3
3515 GREENPOINT AVE, 11101             3
1152 UTICA AVE, 11203                  3
109 E 138TH ST, 10451                  3
355 1ST AVE, 10010                     3
8404 PARSONS BLVD, 11432               3
3500 CONNER ST, 10475                  3
8505 ASTORIA BLVD, 11370               3
Name: count, dtype: int64

In [23]:
# Check uniqueness in weights and measures dataset
wm_unique = wm_data["Inspection Number"].is_unique
print(f"Inspection Numbers unique in weights_measures dataset? {wm_unique}")

Inspection Numbers unique in weights_measures dataset? False


In [24]:
## Now we're looking to cross-reference inspection numbers with charges by merging both datasets, but we must account for duplicate inspection numbers

# Step 1: Load the datasets
wm_data = pd.read_excel("Weights-Measures_Data.xlsx")
charges_data = pd.read_csv("DCWP_Charges.csv")

# Step 2: Filter to failed inspections only (any device condemned in the inspection)
failed_inspections = wm_data[wm_data["Devices Condemned/Confiscated"] > 0].copy()

# Step 3: Clean and align formats
failed_inspections["Inspection Number"] = failed_inspections["Inspection Number"].astype(str)
charges_data["Inspection Number"] = charges_data["Inspection Number"].astype(str)

# Step 4: Aggregate failed inspections to unique inspections (one row per inspection)
failed_inspections_agg = (
    failed_inspections
    .groupby("Inspection Number", as_index=False)
    .agg({
        "Devices Condemned/Confiscated": "sum",
        # add other aggregations if needed (e.g. business name, date by min or max)
    })
)

# Step 5: Count number of charges per inspection
charge_counts = charges_data["Inspection Number"].value_counts().rename("Charge Count")

# Step 6: Merge charge counts into aggregated failed inspections
merged = failed_inspections_agg.merge(charge_counts, how="left", left_on="Inspection Number", right_index=True)

# Step 7: Fill missing charge counts (no charges filed)
merged["Charge Count"] = merged["Charge Count"].fillna(0).astype(int)

# Step 8: Add Charged flag (True if Charge Count > 0)
merged["Charged"] = merged["Charge Count"] > 0

# Step 9: Analyze results
charged_count = merged["Charged"].sum()
total_failed = len(merged)

print(f"{charged_count} of {total_failed} failed inspections resulted in a formal charge.")

# Optional: Print distribution of charge counts
print("\nCharge count distribution:")
print(merged["Charge Count"].value_counts().sort_index())

# Step 10: Filter to only failed inspections that resulted in formal charges
true_charges = merged[merged["Charged"]]

# View the first few rows of the true charges dataset
print(true_charges.head())

# ✅ Step 9: Pull full company info for inspections that were charged
charged_inspection_ids = merged.loc[merged["Charged"], "Inspection Number"]

# Filter original failed_inspections (with full business info) to those IDs
full_info_charged = failed_inspections[failed_inspections["Inspection Number"].isin(charged_inspection_ids)]

# Step 10: Export full dataset of failed inspections that led to charges
full_info_charged.to_csv("Full_Failed_Inspections_With_Charges.csv", index=False)


2 of 1391 failed inspections resulted in a formal charge.

Charge count distribution:
Charge Count
0    1389
1       1
2       1
Name: count, dtype: int64
    Inspection Number  Devices Condemned/Confiscated  Charge Count  Charged
354     INS-000695112                              4             2     True
361     INS-000696027                              2             1     True


In [25]:
## Now pull the full merged dataset
full_info_all = failed_inspections.merge(
    merged[["Inspection Number", "Charge Count", "Charged"]],
    on="Inspection Number",
    how="left"
)

# Step 12: Export full dataset of failed inspections with charge info
full_info_all.to_csv("All_Failed_Inspections_Charges_Flagged.csv", index=False)

It has been reported in the Mayor’s Management Report that the Weights and Measures compliance rate for gasoline pumps reached 100% in fiscal year 2024. However, the code below filters inspection records to focus exclusively on gasoline pumps, removes duplicate inspections to prevent double counting, assigns inspections to fiscal years based on their dates, and calculates the actual compliance rate—the share of pumps passing inspection—by fiscal year. Contrary to the official claim, the calculated compliance rate for fiscal year 2024 is only 69.61%.

In [27]:
import pandas as pd

# Load the Excel file
wm_data = pd.read_excel("Weights-Measures_Data.xlsx")

# Optional: Build full address (for location analysis)
wm_data["Full Address"] = (
    wm_data["Building Number"].fillna('').astype(str).str.strip() + " " +
    wm_data["Street Name"].fillna('').astype(str).str.strip() + ", " +
    wm_data["Postcode"].astype(str)
)

# Filter only 'Gas Pumps' from Device Class
gas_pump_data = wm_data[wm_data["Device Class"] == "Gas Pumps"].copy()
print(f"Gas pump inspection records found: {len(gas_pump_data)}")

# Drop duplicate Inspection Numbers to avoid double-counting
dedup_cols = [
    "Inspection Number", "Date of Occurance",
    "Devices Inspected", "Devices Approved", "Devices Condemned/Confiscated"
]
gas_pump_data_dedup = gas_pump_data[dedup_cols].drop_duplicates(subset=["Inspection Number"])

# Convert date column to datetime format
gas_pump_data_dedup["Date of Occurance"] = pd.to_datetime(gas_pump_data_dedup["Date of Occurance"])

# Add Fiscal Year (NYC Fiscal Year: July 1 – June 30)
gas_pump_data_dedup["Fiscal Year"] = gas_pump_data_dedup["Date of Occurance"].apply(
    lambda x: x.year + 1 if x.month >= 7 else x.year
)

# Group by Fiscal Year and summarize inspections
fy_summary = gas_pump_data_dedup.groupby("Fiscal Year")[[
    "Devices Inspected", "Devices Approved", "Devices Condemned/Confiscated"
]].sum()

# Calculate compliance rate as percentage
fy_summary["Compliance Rate (%)"] = (
    fy_summary["Devices Approved"] / fy_summary["Devices Inspected"]
) * 100

# Round for clarity
fy_summary = fy_summary.round({"Compliance Rate (%)": 2})

# Output summary
print(fy_summary)

Gas pump inspection records found: 3476
             Devices Inspected  Devices Approved  \
Fiscal Year                                        
2024                      3465              2412   
2025                      3645              3426   

             Devices Condemned/Confiscated  Compliance Rate (%)  
Fiscal Year                                                      
2024                                   947                69.61  
2025                                   137                93.99  


In [28]:
wm_data["Date of Occurance"] = pd.to_datetime(wm_data["Date of Occurance"])
wm_data["Fiscal Year"] = wm_data["Date of Occurance"].apply(lambda x: x.year + 1 if x.month >= 7 else x.year)
print("All Fiscal Years Present:", wm_data["Fiscal Year"].unique())

All Fiscal Years Present: [2024 2025]


In [29]:
## Now let's do the same for Fuel Oil Meter Tests
import pandas as pd

# Load the Excel file
wm_data = pd.read_excel("Weights-Measures_Data.xlsx")

# Optional: Build full address (for location analysis)
wm_data["Full Address"] = (
    wm_data["Building Number"].fillna('').astype(str).str.strip() + " " +
    wm_data["Street Name"].fillna('').astype(str).str.strip() + ", " +
    wm_data["Postcode"].astype(str)
)

# Filter only 'Fuel Oil Meter Test' from Device Class
fuel_meter_data = wm_data[wm_data["Device Class"] == "Fuel Oil Meter Test"].copy()
print(f"Fuel Oil Meter inspection records found: {len(fuel_meter_data)}")

# Drop duplicate Inspection Numbers to avoid double-counting
dedup_cols = [
    "Inspection Number", "Date of Occurance",
    "Devices Inspected", "Devices Approved", "Devices Condemned/Confiscated"
]
fuel_meter_data_dedup = fuel_meter_data[dedup_cols].drop_duplicates(subset=["Inspection Number"])

# Convert date column to datetime format
fuel_meter_data_dedup["Date of Occurance"] = pd.to_datetime(fuel_meter_data_dedup["Date of Occurance"])

# Add Fiscal Year (NYC Fiscal Year: July 1 – June 30)
fuel_meter_data_dedup["Fiscal Year"] = fuel_meter_data_dedup["Date of Occurance"].apply(
    lambda x: x.year + 1 if x.month >= 7 else x.year
)

# Group by Fiscal Year and summarize inspections
fy_summary = fuel_meter_data_dedup.groupby("Fiscal Year")[[
    "Devices Inspected", "Devices Approved", "Devices Condemned/Confiscated"
]].sum()

# Calculate compliance rate as percentage
fy_summary["Compliance Rate (%)"] = (
    fy_summary["Devices Approved"] / fy_summary["Devices Inspected"]
) * 100

# Round for clarity
fy_summary = fy_summary.round({"Compliance Rate (%)": 2})

# Output summary
print(fy_summary)

Fuel Oil Meter inspection records found: 997
             Devices Inspected  Devices Approved  \
Fiscal Year                                        
2024                       591               512   
2025                       403               365   

             Devices Condemned/Confiscated  Compliance Rate (%)  
Fiscal Year                                                      
2024                                   153                86.63  
2025                                    54                90.57  


Since the Business Unique ID and Business Name are a common column value in the Weights and Measures Inspection Data Set, DCWP Charges Data Set, and DCWP Consumer Complaints, we will analyze each dataset, the top 10 most common repeat offenders for each dataset and then compare out of the Weights and Measures Inspection, how often the top offenders appear in the DCWP Charges Data Set and DCWP Consumer Complaints. We want to remove the trailing year suffix (e.g., -2025) for the Business Unique ID before grouping; otherwise, businesses will be split into different IDs even though they’re the same. 

In [31]:
#Let's start with the Weights and Measures inspection data. Because there are multiple devices for a single inspection event, we have to remove duplicates.

import pandas as pd

# Load the data
wm_data = pd.read_excel("Weights-Measures_Data.xlsx")

# Remove duplicate entries for the same inspection event
wm_data_clean = wm_data.drop_duplicates(subset='Inspection Number')

# Remove year suffix from Business Unique ID
wm_data_clean['Base Business ID'] = wm_data_clean['Business Unique ID'].str.replace(r'-\d{4,5}$', '', regex=True)

# Group by normalized Business ID and Business Name
inspection_counts = (
    wm_data_clean
    .groupby(['Base Business ID', 'Business Name'])
    .size()
    .reset_index(name='Unique Inspection Count')
    .sort_values(by='Unique Inspection Count', ascending=False)
)

# Top 10 businesses with the most unique inspections
top_10_offenders = inspection_counts.head(10)

# Display result
top_10_offenders

/var/folders/6p/9zpd5gjx1dj6zpc834rhv6d40000gn/T/ipykernel_45694/2589053644.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wm_data_clean['Base Business ID'] = wm_data_clean['Business Unique ID'].str.replace(r'-\d{4,5}$', '', regex=True)


,Base Business ID,Business Name,Unique Inspection Count
4028,BA-1545133,SPRAGUE OPERATING RESOURCES LLC,136
453,BA-1105443,APPROVED OIL CO OF BROOKLYN INC,98
2995,BA-1422612,PETRO INC.,63
4150,BA-1551335,STAR ENERGY TRANSPORTATION INC.,50
4137,BA-1550342,UNITED APOLLO PETROLEUM TRANSPORTATION CORP,46
2702,BA-1370101,JAG TRANSPORTING INC,36
1085,BA-1130517,ATLANTIC COAST TRANSPORT GROUP INC,35
4113,BA-1549432,UNITED FUEL SERVICE LLC.,34
1851,BA-1260245,ENERGY CONSERVATION GROUP LLC,31
5643,BA-1676415,DEPENDABLE TRANSPORT INC,25


In [62]:
## Now let's do the DCWP Charges

import pandas as pd

# Load and clean data
charges_data = pd.read_csv("DCWP_Charges.csv")

# Remove duplicate Record ID + Inspection Number combinations
charges_data_cleaned = charges_data.drop_duplicates(subset=['Record ID', 'Inspection Number'])

# Remove the year suffix from Business Unique ID (e.g., -2025)
charges_data_cleaned['Base Business ID'] = charges_data_cleaned['Business Unique ID'].str.replace(r'-\d{4,5}$', '', regex=True)

# Group by Business Name and normalized Business ID
top_offenders = (
    charges_data_cleaned
    .groupby(['Business Name', 'Base Business ID'])
    .size()
    .reset_index(name='Charge Count')
    .sort_values(by='Charge Count', ascending=False)
    .head(10)
)

print(top_offenders)

                             Business Name Base Business ID  Charge Count
11949                   PATEL, JYOTINDRA A       BA-1406409            22
6941      GET ONE CONVENIENCE & MORE CORP.       BA-1683448            20
995                  3029 CONVENIENT CORP.       BA-1680875            19
1000              304 ORGANIC MARKET CORP.       BA-1681294            18
3347              BENTLEY CONVENIENCE CORP       BA-1118852            18
2619   AMIGOS GROCERY AND CANDY STORE CORP       BA-1125623            17
13679                 SMOKE ZONE PLUS CORP       BA-1552310            16
13677                    SMOKE ZONE 2, INC       BA-1532654            16
10148               MAZAJ SMOKE PLACE CORP       BA-1571795            16
13695             SMOKEYS CONVENIENCE CORP       BA-1575824            16


/var/folders/6p/9zpd5gjx1dj6zpc834rhv6d40000gn/T/ipykernel_45694/1582432404.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  charges_data_cleaned['Base Business ID'] = charges_data_cleaned['Business Unique ID'].str.replace(r'-\d{4,5}$', '', regex=True)


In [70]:
#Now let's do the DCWP complaints 

import pandas as pd

# Load the dataset
complaints_data = pd.read_csv("DCWP_Consumer_Complaints.csv")

# Check dataset info
complaints_data.info()

# Check for duplicate Record IDs
duplicate_record_ids = complaints_data.duplicated(subset='Record ID').sum()
print(f"Duplicate Record IDs: {duplicate_record_ids}")

# Remove duplicate Record IDs
complaints_data_cleaned = complaints_data.drop_duplicates(subset='Record ID')

# Truncate last five digits of Business Unique ID (if it's a string or int)
complaints_data_cleaned['Business Unique ID'] = complaints_data_cleaned['Business Unique ID'].astype(str).str[:-5]

# Group by Business Name and truncated Business Unique ID, count complaints
top_complaint_offenders = (
    complaints_data_cleaned
    .groupby(['Business Name', 'Business Unique ID'])
    .size()
    .reset_index(name='Complaint Count')
    .sort_values(by='Complaint Count', ascending=False)
    .head(10)
)

# Display top 10 offenders
print(top_complaint_offenders)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45509 entries, 0 to 45508
Data columns (total 33 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Record ID                  45509 non-null  object 
 1   Intake Date                45509 non-null  object 
 2   Intake Channel             45509 non-null  object 
 3   311 SR Number              33354 non-null  object 
 4   Business Category          44138 non-null  object 
 5   Complaint Code             38064 non-null  object 
 6   Business Unique ID         43675 non-null  object 
 7   Business Name              43288 non-null  object 
 8   Result Date                45509 non-null  object 
 9   Result                     45509 non-null  object 
 10  Referred To                18167 non-null  object 
 11  Contract Cancelled Amount  35 non-null     float64
 12  Refund Amount              938 non-null    float64
 13  Address Type               42158 non-null  obj

/var/folders/6p/9zpd5gjx1dj6zpc834rhv6d40000gn/T/ipykernel_45694/2699294252.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  complaints_data_cleaned['Business Unique ID'] = complaints_data_cleaned['Business Unique ID'].astype(str).str[:-5]


In [72]:
#Now let's see how often the W&M Inspection top repeat offenders show up across the datasets

import pandas as pd

# Load both datasets
charges_df = pd.read_csv("DCWP_Charges.csv")
complaints_df = pd.read_csv("DCWP_Consumer_Complaints.csv")

# Ensure Business Unique ID is treated as string and truncate last 5 characters
charges_df['Business Unique ID'] = charges_df['Business Unique ID'].astype(str).str[:-5]
complaints_df['Business Unique ID'] = complaints_df['Business Unique ID'].astype(str).str[:-5]

# List of businesses with truncated Business Unique ID
business_list = [
    ("BA-1545133", "SPRAGUE OPERATING RESOURCES LLC"),
    ("BA-1105443", "APPROVED OIL CO OF BROOKLYN INC"),
    ("BA-1422612", "PETRO INC."),
    ("BA-1551335", "STAR ENERGY TRANSPORTATION INC."),
    ("BA-1550342", "UNITED APOLLO PETROLEUM TRANSPORTATION CORP"),
    ("BA-1370101", "JAG TRANSPORTING INC"),
    ("BA-1130517", "ATLANTIC COAST TRANSPORT GROUP INC"),
    ("BA-1549432", "UNITED FUEL SERVICE LLC."),
    ("BA-1260245", "ENERGY CONSERVATION GROUP LLC"),
    ("BA-1676415", "DEPENDABLE TRANSPORT INC")
]

# Initialize results list
results = []

# Count appearances in both datasets
for bid, name in business_list:
    charges_count = charges_df[
        (charges_df['Business Unique ID'] == bid) &
        (charges_df['Business Name'].str.upper() == name.upper())
    ].shape[0]
    
    complaints_count = complaints_df[
        (complaints_df['Business Unique ID'] == bid) &
        (complaints_df['Business Name'].str.upper() == name.upper())
    ].shape[0]
    
    results.append((bid, name, charges_count, complaints_count))

# Convert to DataFrame
results_df = pd.DataFrame(results, columns=[
    "Business Unique ID", "Business Name", "Charges Appearances", "Complaints Appearances"
])

# Display the results
print(results_df)

  Business Unique ID                                Business Name  \
0         BA-1545133              SPRAGUE OPERATING RESOURCES LLC   
1         BA-1105443              APPROVED OIL CO OF BROOKLYN INC   
2         BA-1422612                                   PETRO INC.   
3         BA-1551335              STAR ENERGY TRANSPORTATION INC.   
4         BA-1550342  UNITED APOLLO PETROLEUM TRANSPORTATION CORP   
5         BA-1370101                         JAG TRANSPORTING INC   
6         BA-1130517           ATLANTIC COAST TRANSPORT GROUP INC   
7         BA-1549432                     UNITED FUEL SERVICE LLC.   
8         BA-1260245                ENERGY CONSERVATION GROUP LLC   
9         BA-1676415                     DEPENDABLE TRANSPORT INC   

   Charges Appearances  Complaints Appearances  
0                    1                       0  
1                    0                       1  
2                    1                       1  
3                    1                      

In [74]:
# Now let's examine the fines and fees dataset

# Load the dataset
fines_fees = pd.read_csv("DCWP_Fines_and_Fees__Historical.csv") 

fines_fees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 252556 entries, 0 to 252555
Data columns (total 11 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   RECORD ID        252556 non-null  object 
 1   RECORD TYPE      252556 non-null  object 
 2   BUSINESS NAME    251767 non-null  object 
 3   BUSINESS NAME2   21983 non-null   object 
 4   INDUSTRY         252407 non-null  object 
 5   FEE SEQUENCE ID  252556 non-null  int64  
 6   FEE TYPE         252556 non-null  object 
 7   FEE DESCRIPTION  252556 non-null  object 
 8   FEE AMOUNT       252556 non-null  float64
 9   FEE DATE         252556 non-null  object 
 10  FEE STATUS       252556 non-null  object 
dtypes: float64(1), int64(1), object(9)
memory usage: 21.2+ MB


In [76]:
sorted(fines_fees['FEE DESCRIPTION'].dropna().unique())

["'TO - Tobacco Other",
 'Additional License Fee for Las Vegas Games (GC-7)',
 'Additional License Fee for Raffles (GC-7R)',
 'Additional Pedicab',
 'Amendment: Per Occasion Fee for Las Vegas Night Games',
 'Amusement Arcade License Fee',
 'Amusement Device (Temporary) DOB Mechanical Device Inspection Fee',
 'Amusement Device (Temporary) License Fee',
 'Annual Fee for Bell Jar Games',
 'Annual Fee for Raffle Games',
 'Appeals - Consumer Restitution Collected',
 'Auction House Premises License Blue Dot Fee',
 'Auction House Premises License Fee',
 'Auction House Premises License Renewal Fee',
 'Auctioneer Blue Dot Fee',
 'Auctioneer License Fee',
 'Auctioneer Renewal Fee',
 'BC-6 Amendment - Additional Per Occasion Fee',
 'BC-7 Additional License Fee',
 'Blue Dot Fee',
 'Blue Dot Fee for Employment Agency',
 'Blue Dot Fee for Tow Truck Company',
 'Bluedot Fee',
 'Booting Company License Blue Dot Fee',
 'Booting Company License Fee',
 'Booting Company License Renewal Fee',
 'CD - Consume

In [92]:
import pandas as pd
from difflib import get_close_matches

# 1. Load datasets
fines_fees = pd.read_csv("DCWP_Fines_and_Fees__Historical.csv")
inspections = pd.read_excel("Weights-Measures_Data.xlsx")  # Replace with your actual filename

# 2. Standardize all column names to uppercase with no leading/trailing spaces
fines_fees.columns = fines_fees.columns.str.strip().str.upper()
inspections.columns = inspections.columns.str.strip().str.upper()

# 3. Clean business names
fines_fees['BUSINESS NAME'] = fines_fees['BUSINESS NAME'].astype(str).str.upper().str.strip()
inspections['BUSINESS NAME'] = inspections['BUSINESS NAME'].astype(str).str.upper().str.strip()

# 4. Extract top offenders from inspections dataset
top_offenders = (
    inspections['BUSINESS NAME']
    .value_counts()
    .head(10)  # Adjust this number if needed
    .index
    .tolist()
)

# 5. Define exact W&M-related fee descriptions to filter by
wm_keywords = [
    'LIQUID TEST MEASURE',
    'PETROL METER TYPE A',
    'PETROL METER TYPE B',
    'PETROL PUMP BLEND',
    'PETROL PUMP DIESEL',
    'PETROL PUMP DUAL',
    'PETROL PUMP SINGLE',
    'SCALE TO 3,307 LBS',
    'SCALE TO 33 LBS',
    'SCALE TO 661 LBS',
    'SCALE DEALER REPAIRER LICENSE',
    'WH - W&M HEARABLE VIOLATION',
    'WM - W&M VIOLATION',
    'WS - W&H NON-HEARABLE VIOLATION',
    'TRUCK WEIGHT TEST',
    'TIMING DEVICE',
]

# Build regex pattern from keywords (escaped to avoid regex special chars)
import re
pattern = '|'.join([re.escape(kw) for kw in wm_keywords])

# 6. Filter fines_fees for only W&M related fees based on exact descriptions
wm_fees = fines_fees[
    fines_fees['FEE DESCRIPTION'].astype(str).str.upper().str.contains(pattern, na=False)
]

# 7. Unique business names in fines/fees dataset filtered for W&M
fine_businesses = wm_fees['BUSINESS NAME'].dropna().unique().tolist()

# 8. Fuzzy match top offenders to fines/fees dataset using difflib
results = []

for offender in top_offenders:
    match_list = get_close_matches(offender, fine_businesses, n=1, cutoff=0.6)
    if match_list:
        match = match_list[0]
        matched_rows = wm_fees[wm_fees['BUSINESS NAME'] == match]
        results.append({
            'Offender Name (Inspections)': offender,
            'Matched Name (Fines/Fees)': match,
            'Number of W&M-related Fees': len(matched_rows)
        })
    else:
        results.append({
            'Offender Name (Inspections)': offender,
            'Matched Name (Fines/Fees)': None,
            'Number of W&M-related Fees': 0
        })

# 9. Convert results to DataFrame and sort descending by fees count
results_df = pd.DataFrame(results)
results_df = results_df.sort_values(by='Number of W&M-related Fees', ascending=False)

# 10. Display results
print(results_df)

                   Offender Name (Inspections)  \
0                        BOLLA OPERATING CORP.   
3              APPROVED OIL CO OF BROOKLYN INC   
1              SPRAGUE OPERATING RESOURCES LLC   
5                                  PETRO  INC.   
2                                 SPEEDWAY LLC   
7              STAR ENERGY TRANSPORTATION INC.   
8  UNITED APOLLO PETROLEUM TRANSPORTATION CORP   
4      THE STOP & SHOP SUPERMARKET COMPANY LLC   
9                                 NAMDOR, INC.   
6                             AMG RETAIL I LLC   

                     Matched Name (Fines/Fees)  Number of W&M-related Fees  
0                        BOLLA OPERATING CORP.                         231  
3              APPROVED OIL CO OF BROOKLYN INC                         180  
1              SPRAGUE OPERATING RESOURCES LLC                         179  
5                                   PETRO INC.                         155  
2                                 SPEEDWAY LLC                  

In [94]:
import pandas as pd
import re

# Load the dataset
fines_fees = pd.read_csv("DCWP_Fines_and_Fees__Historical.csv")

# Standardize column names
fines_fees.columns = fines_fees.columns.str.strip().str.upper()

# Define exact W&M-related fee descriptions to filter by (all uppercase for consistent matching)
wm_keywords = [
    'LIQUID TEST MEASURE',
    'PETROL METER TYPE A',
    'PETROL METER TYPE B',
    'PETROL PUMP BLEND',
    'PETROL PUMP DIESEL',
    'PETROL PUMP DUAL',
    'PETROL PUMP SINGLE',
    'SCALE TO 3,307 LBS',
    'SCALE TO 33 LBS',
    'SCALE TO 661 LBS',
    'SCALE DEALER REPAIRER LICENSE',
    'WH - W&M HEARABLE VIOLATION',
    'WM - W&M VIOLATION',
    'WS - W&H NON-HEARABLE VIOLATION',
    'TRUCK WEIGHT TEST',
    'TIMING DEVICE',
]

# Build regex pattern for exact matching
pattern = '|'.join([re.escape(kw) for kw in wm_keywords])

# Filter for W&M-related fees using exact fee descriptions, case insensitive
wm_fees = fines_fees[
    fines_fees['FEE DESCRIPTION'].astype(str).str.upper().str.contains(pattern, na=False)
]

# Group by business name and count number of W&M-related fees
wm_top10 = (
    wm_fees.groupby('BUSINESS NAME')
    .size()
    .reset_index(name='NUMBER OF W&M-RELATED FEES')
    .sort_values(by='NUMBER OF W&M-RELATED FEES', ascending=False)
    .head(10)
)

print(wm_top10)

                                    BUSINESS NAME  NUMBER OF W&M-RELATED FEES
1931                        BOLLA OPERATING CORP.                         231
1453              APPROVED OIL CO OF BROOKLYN INC                         180
7400              SPRAGUE OPERATING RESOURCES LLC                         179
6495                                   PETRO INC.                         155
7390                                 SPEEDWAY LLC                         141
7433              STAR ENERGY TRANSPORTATION INC.                          74
7959  UNITED APOLLO PETROLEUM TRANSPORTATION CORP                          68
7776      THE STOP & SHOP SUPERMARKET COMPANY LLC                          64
7971                     UNLIMITED TRUCKING CORP.                          50
8212                 WHOLE FOODS MARKET GROUP INC                          48
